dbt/adapters/duckdb/impl.py

```diff
@@ -1,14 +1,52 @@
-from dbt.adapters.duckdb import DuckDBConnectionManager
+from dbt.adapters.duckdb.connections import DuckDBConnectionManager
 from dbt.adapters.sql import SQLAdapter
+from dbt.contracts.connection import AdapterResponse
+from dbt.exceptions import InternalException, RuntimeException
 
 
 class DuckDBAdapter(SQLAdapter):
     ConnectionManager = DuckDBConnectionManager
 
     @classmethod
-    def date_function(cls):
+    def date_function(cls) -> str:
         return "now()"
 
     @classmethod
-    def is_cancelable(cls):
+    def is_cancelable(cls) -> bool:
         return False
+
+    def valid_incremental_strategies(self):
+        """The set of standard builtin strategies which this adapter supports out-of-the-box.
+        Not used to validate custom strategies defined by end users.
+        """
+        return ["append", "delete+insert", "merge", "insert_overwrite"]
+
+    def commit_if_has_connection(self) -> None:
+        """This is just a quick-fix. Python models do not execute begin function so the transaction_open is always false."""
+        try:
+            self.connections.commit_if_has_connection()
+        except InternalException:
+            pass
+
+    def submit_python_job(
+        self, parsed_model: dict, compiled_code: str
+    ) -> AdapterResponse:
+
+        connection = self.connections.get_if_exists()
+        if not connection:
+            connection = self.connections.get_thread_connection()
+        con = connection.handle._conn
+
+        def load_df_function(table_name: str):
+            """Currently con.table method dos not support fully qualified name - https://github.com/duckdb/duckdb/issues/5038"""
+            return con.query(f"select * from {table_name}")
+
+        try:
+            exec(compiled_code)
```

Comment:
FYI that this approach very much resonates with me, but I sort of suspect that we should do some work to limit the scope of the `exec` call by specifying the `globals` and `locals` dictionaries, right? Like we would e.g. not want the script running to be able to access/modify the state of the dbt run beyond whatever scope we passed into it.

Reply:
Honestly, I was not thinking about it at all :). It is a really great point. I have limited both dictionaries to a bare minimum. `load_df_function` can be deleted in the future after https://github.com/duckdb/duckdb/issues/5038 is fixed.